bob/src/cluster/quorum.rs

```diff
@@ -318,6 +319,91 @@ impl Quorum {
         }
         Ok(())
     }
+
+    async fn collect_remote_exists(
+        result: &mut [bool],
+        keys: &[BobKey],
+        indexes_by_node: &mut HashMap<Node, IndexMap>,
+    ) -> Result<(), Error> {
+        if !indexes_by_node.is_empty() {
+            let mut keys_by_node = HashMap::new();
+            for (node, node_map) in indexes_by_node.iter_mut() {
+                node_map.retain_not_existed(&result);
+                if !node_map.is_empty() {
+                    keys_by_node.insert(node.clone(), node_map.collect(keys));
+                }
+            }
+
+            if !keys_by_node.is_empty() {
+                let nodes: Vec<_> = keys_by_node.keys().collect();
+                let remote_results = exist_on_remote_nodes(&nodes, &keys_by_node).await;
+                for (remote_result, node) in remote_results.into_iter().zip(nodes) {
```

Comment:
Unfortunately, this `zip` is incorrect, because `exist_on_remote_nodes` calls `LinkManager::call_nodes` which uses `FuturesUnordered<_>` inside. Thus the order of the results is not preserved. 
The mapping should use node name stored within `remote_result`.

Possible solution: add `nodes_by_name: HashMap<String, Node>` to `Virtual` inside `mapper.rs` and function `fn node_by_name(&self, name: &str) -> Option<&Node>` to get node by its name. Then here you can pass only `keys_by_node` to `exist_on_remote_nodes` (`nodes` vector is not needed anymore). And to get the `IndexMap` you can use double indirection `keys_by_node.get(mapper.node_by_name(remote_result.node_name()))`